sessions/tracker.go

```diff
@@ -0,0 +1,94 @@
+package sessions
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+const (
+	//contextSessionKey is a unique key for accessing and setting Bugsnag
+	//session data on a context.Context object
+	contextSessionKey ctxKey = 1
+)
+
+// ctxKey is a type alias that ensures uniqueness as a context.Context key
+type ctxKey int
+
+// SessionTracker exposes a method for starting sessions that are used for
+// gauging your application's health
+type SessionTracker interface {
+	StartSession(context.Context) context.Context
+}
+
+type sessionTracker struct {
+	sessionChannel chan *session
+	sessions       []*session
+	config         *SessionTrackingConfiguration
+	publisher      sessionPublisher
+}
+
+// NewSessionTracker creates a new SessionTracker based on the provided config,
+func NewSessionTracker(config *SessionTrackingConfiguration) SessionTracker {
+	publisher := publisher{
+		config: config,
+		client: &http.Client{Transport: config.Transport},
+	}
+	st := sessionTracker{
+		sessionChannel: make(chan *session, 1),
+		sessions:       []*session{},
+		config:         config,
+		publisher:      &publisher,
+	}
+	go st.processSessions()
+	return &st
+}
+
+func (s *sessionTracker) StartSession(ctx context.Context) context.Context {
+	session := newSession()
+	s.sessionChannel <- session
+	return context.WithValue(ctx, contextSessionKey, session)
+}
+
+func (s *sessionTracker) interval() time.Duration {
+	s.config.mutex.Lock()
```

Comment:
Is there any other places where we need to lock the config? Looking at the session tracker and publisher structs, they seem to access the config often. Could this be modified underneath them by the user?

Reply:
I've added a lock around publishing. I had a look around the sessions package and I think this is the only location where we need to worry about a race condition.